README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# linemap: Create maps made of Lines <img src="man/figures/logo.png" align="right" width="140"/>

[![R-CMD-check](https://github.com/riatelab/linemap/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/riatelab/linemap/actions/workflows/R-CMD-check.yaml)
[![Codecov test
coverage](https://codecov.io/gh/riatelab/linemap/branch/master/graph/badge.svg)](https://app.codecov.io/gh/riatelab/linemap?branch=master)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version-ago/linemap)](https://cran.r-project.org/package=linemap)

`linemap()` displays a map made of lines using a raster or gridded data.

## Installation

You can install `linemap` from CRAN with:

``` r
install.packages("linemap")
```

## Example

The package contains one function: `linemap()`.

``` r
library(linemap)
r <- terra::rast(system.file("tif/elevation.tif", package = "linemap"))
opar <- par(mar = c(0,0,0,0), bg = "grey10")
linemap(r)
par(opar)
```

![](man/figures/README-example-1.png)<!-- -->

## Inspiration

[Unknown Pleasures](https://en.wikipedia.org/wiki/Unknown_Pleasures)
(*Joy Division*)  
[Population Lines: How and Why I Created
It](https://jcheshire.com/featured-maps/population-lines-how-and-why-i-created-it/)
(*James Cheshire*)  
[GeospatialLineGraphs](https://github.com/Brideau/GeospatialLineGraphs)
(*Ryan Brideau*)

## Alternative

- ggplot2 + [ggridges](https://CRAN.R-project.org/package=ggridges)


NEWS.md
# linemap 0.3.0

## Major change

- Use terra raster as input
- remove former example dataset


# linemap 0.2.0

## Minor changes
- Remove some RData in favor of geopackage (This can lead to breaking changes using examples)
- fix NOTES by importing geopackage instead of RData
- Add test, code coverage and GH action
- Fix links



tests/tinytest.R
if ( requireNamespace("tinytest", quietly=TRUE) ){
  tinytest::test_package("linemap")
}


inst/tinytest/test_linemap.R
library(terra)
r <- rast(system.file("tif/elevation.tif", package = "linemap"))

expect_silent(linemap(x = r, k = 2.5, threshold = 50,
                      col = "ivory1", border = "ivory4",
                      lwd = 0.6,
                      add = FALSE))

expect_error(linemap(x = 8))
s <- rast(system.file("ex/logo.tif", package = "terra"))
expect_error(linemap(x = s))

pts <- as.points(r)
mon_df <- data.frame(crds(pts), values(pts))
expect_silent(linemap(x = mon_df))
expect_error(linemap(mon_df[1:130, 1:2]))
expect_error(linemap(mon_df[c(1:1200,3,5), ]))
mon_df$x = "aha"
expect_error(linemap(mon_df))


R/getgrid.R
#' @title Transform a Polygon Layer to a Grid
#' @name getgrid
#' @description Transform a polygon layer to a regular grid data.frame.
#' @param x an sf polygon layer.
#' @param cellsize size of the side of a grid cell.
#' @param var name of the variable to transform to the grid. It can be a vector
#' of names.
#' @return A data frame is returned.
#' @export
#' @examples
#' library(linemap)
#' library(sf)
#' Bretagne <- st_read(system.file("gpkg/geofla.gpkg", package = "linemap"),
#'                     layer = "Bretagne")
#' France <- st_read(system.file("gpkg/geofla.gpkg", package = "linemap"),
#'                   layer = "France")
#' # example on an extract of dataset
#' cotedarmor <- Bretagne[Bretagne$CODE_DEPT == 22, ]
#' cota <- getgrid(x = cotedarmor, cellsize = 1750, var = "POPULATION")
#' opar <- par(mar = c(0,0,0,0))
#' plot(st_geometry(France), col="lightblue3", border = NA, bg = "lightblue2",
#'      xlim = c(min(cota$X), max(cota$X)), ylim= c(min(cota$Y), max(cota$Y)))
#' linemap(x = cota, var = "POPULATION", k = 5, threshold = 1,
#'         col = "lightblue3", border = "white", lwd = 0.8,
#'         add = TRUE)
#' par(opar)
#'
#'
#' \donttest{
#' # example on the full dataset
#' Bretagne_grid <- getgrid(x = Bretagne, cellsize = 1750, var = "POPULATION")
#' opar <- par(mar = c(0,0,0,0))
#' plot(st_geometry(France), col="lightblue3", border = NA, bg = "lightblue2",
#'      xlim = range(Bretagne_grid$X), ylim= range(Bretagne_grid$Y))
#' linemap(x = Bretagne_grid, var = "POPULATION", k = 5, threshold = 1,
#'         col = "lightblue3", border = "white", lwd = 0.8,
#'         add = TRUE)
#' par(opar)
#' }
getgrid <- function(x, cellsize, var){
  b <- sf::st_bbox(x)
  mx <- (b[3] - b[1]) %/% cellsize
  my <- (b[4] - b[2]) %/% cellsize
  if((b[3] - b[1]) %% cellsize!=0){
    b[3] <- b[1] + (mx+1) * cellsize
  }
  if((b[4] - b[2]) %% cellsize!=0){
    b[4] <- b[2] + (my+1) * cellsize
  }
  n <- unname(c(ceiling(diff(b[c(1, 3)]) / cellsize),
                ceiling(diff(b[c(2, 4)]) / cellsize)))
  grid <- sf::st_make_grid(cellsize = cellsize, offset = b[1:2],
                       n = n,crs = sf::st_crs(x))
  grid <- sf::st_sf(id_cell=1:length(grid), geometry = grid)
  row.names(grid) <- grid$id_cell

  x$area <- sf::st_area(x)
  # predicted warning, we don't care...
  # options(warn = -1)
  sf::st_agr(x) <- 'constant'
  sf::st_agr(grid) <- 'constant'

  parts <- sf::st_intersection(x = grid[,"id_cell"], y = x)
  # options(warn = 0)

  parts$area_part <- sf::st_area(parts)
  lvar <- vector(mode = "list", length = length(var))
  names(lvar) <- var
  for (i in 1:length(lvar)){
    lvar[[i]] <- as.vector(
      parts[[names(lvar)[i]]] * parts$area_part / parts$area
      )
  }
  v <- stats::aggregate(do.call(cbind,lvar), by = list(id = parts[['id_cell']]),
                 FUN = sum, na.rm=TRUE)
  grid <- merge(grid, v, by.x  = "id_cell", by.y = "id", all.x = T)
  gr <- cbind(grid, sf::st_coordinates(sf::st_centroid(sf::st_geometry(grid))))
  gr <- as.data.frame(gr[,c("X","Y",var), drop=TRUE])
  return(gr)
}






R/linemap.R
#' @title Create Line Map
#' @name linemap
#' @description Plot a line map.
#' @param x a data.frame (x, y, value), a terra SpatRaster.
#' @param k expension factor.
#' @param threshold threshold of the data to plot.
#' @param col color for the lines areas.
#' @param border color for the lines borders.
#' @param lwd thickness of the lines.
#' @param bg background color
#' @param add if TRUE add the lines to the current plot.
#' @importFrom graphics par
#' @export
#' @examples
#' ras <- terra::rast(system.file("tif/elevation.tif", package = "linemap"))
#' linemap(ras, k = 7)
linemap <- function(x,
                    k = 2,
                    threshold = 1,
                    col = "white",
                    border = "black",
                    lwd = 0.5,
                    bg = par("bg"),
                    add = FALSE) {

  x <- input_mat(x)
  mmin <- min(x, na.rm = TRUE)
  x[is.na(x)] <- 0
  x[is.nan(x)] <- 0
  x[is.infinite(x)] <- 0
  x[x < threshold] <- 0
  dx <- dim(x)
  lon <- as.numeric(colnames(x))
  lat <- as.numeric(row.names(x))
  mm <- x * k + lat

  # ausol
  mm <- mm - mmin

  if(!add){
    graphics::plot(1:10, type = "n", axes = F,
                   xlab = "", ylab="", asp = 1,
                   xlim = c(min(lon), max(lon)),
                   ylim = c(min(c(mm, lat)), max(c(mm, lat))))
  }
  long <- c(lon[1], lon, lon[dx[2]])
  ncolm <- dx[2]
  rrowm <- dx[1]
  for (i in seq_along(lat)){
    graphics::polygon(x = long,
                      y = c(lat[i], mm[i,], lat[i]),
                      border = NA,
                      col = col)
    mc <- rep(border, ncolm - 1)
    v1 <- unname(x[i, ] > threshold)
    v <- v1[-ncolm] + v1[-1]
    mc[v == 0] <- NA
    graphics::segments(x0 = lon[-ncolm],
                       y0 = mm[i,][-ncolm],
                       x1 = lon[-1],
                       y1 = mm[i,][-1],
                       col = mc,
                       lwd = lwd
    )
  }
  graphics::polygon(x = long,
          y = c(lat[rrowm], mm[rrowm,], lat[rrowm]),
          border = NA,
          col = bg)
  graphics::segments(x0 = lon[-ncolm],
           y0 = mm[rrowm, ][-ncolm],
           x1 = lon[-1],
           y1 = mm[rrowm, ][-1],
           col = mc,
           lwd = lwd
  )
}



input_mat <- function(x) {
  # test inputs
  if (!inherits(x = x,
                what = c("SpatRaster",
                         "data.frame"))) {
    stop(
      paste0(
        "'x' should be a data.frame ",
        "or a SpatRaster."
      ),
      call. = FALSE
    )
  }

  if (inherits(x = x, what = c("SpatRaster"))) {
    if (terra::nlyr(x) != 1) {
      stop("'x' should be a single layer SpatRaster.",
           call. = FALSE)
    }
    ext <- terra::ext(x)
    nc <- terra::ncol(x)
    nr <- terra::nrow(x)
    xr <- terra::xres(x) / 2
    yr <- terra::yres(x) / 2
    lon <- seq(ext[1] + xr, ext[2] - xr, length.out = nc)
    lat <- seq(ext[4] - yr, ext[3] + yr, length.out = nr)
    m <- matrix(
      data = terra::values(x),
      nrow = nr,
      dimnames = list(lat, lon),
      byrow = TRUE
    )
  }

  if (inherits(x = x, what = "data.frame")) {
    nax <- names(x)
    if (length(nax) != 3) {
      stop("`x` should contains 3 variables : x, y and value.", call. = FALSE)
    }
    if(!all(apply(x, 2, is.numeric))){
      stop("x, y and value should be numeric.", call. = FALSE)
    }

    coords <- nax[1:2]
    var <- nax[3]

    if (length(unique(x[[coords[1]]])) * length(unique(x[[coords[2]]])) !=
        length(x[[var]])) {
      stop("It seems that 'x' is not a regular grid.",
           call. = FALSE)
    }

    # Reorder dataframe by X-Y if needed
    x <- x[order(x[[coords[1]]]), ]
    x <- x[order(x[[coords[2]]], decreasing = TRUE), ]

    m <- t(matrix(
      data = x[[var]],
      nrow = length(unique(x[[coords[1]]])),
      dimnames = list(unique(x[[coords[1]]]),
                      unique(x[[coords[2]]]))
    ))
  }
  return(m)
}


R/openbar.R
#
# load("/home/tg/Documents/prz/semin-r_2016/data/BarsParis.RData")
# library(sf)
#
# getGridSquare <- function(x, cellsize){
#   # cellsize transform
#   cellsize = sqrt(cellsize)
#   boundingBox <- st_bbox(x)
#   rounder <- boundingBox[1:2] %% cellsize
#   boundingBox[1] <- boundingBox[1] - rounder[1]
#   boundingBox[2] <- boundingBox[2] - rounder[2]
#   n <- unname(c(ceiling(diff(boundingBox[c(1, 3)]) / cellsize),
#                 ceiling(diff(boundingBox[c(2, 4)]) / cellsize)))
#   grid <- st_make_grid(cellsize = cellsize, offset = boundingBox[1:2], n = n,
#                        crs = st_crs(x))
#   grid <- st_sf(id_cell=1:length(grid), geometry = grid)
#   # grid$id_cell <- 1:nrow(grid)
#   row.names(grid) <- grid$id_cell
#   return(grid)
# }
#
# bars <- st_as_sf(bars)
# plot(bars$geometry)
# grid <- getGridSquare(x = bars, cellsize = 300 * 300 )
# plot(grid$geometry, add=T)
#
# inter <- st_intersects(x = bars, y = grid)
#
#
# bars$id_cell <- unlist(inter)
#
# grid_bar <- aggregate(bars$id_cell, by = list(id_cell=bars$id_cell), length)
# grid <- merge(grid, grid_bar, by = "id_cell", all.x = TRUE)
#
# names(grid)[2] <- "nbars"
# x <- cbind(grid, st_coordinates(st_centroid(grid)))
# x[is.na(x$nbars),"nbars"] <- 0
#
# x <- as.data.frame(x[c("X", "Y", "nbars")])
#
#
# library(linemap)
# par(mar=c(0,0,0,0), bg = "ivory2")
# plot(paris, col="ivory1", border = NA, lwd = 1, lty=3)
# linemap(x = x, var = "nbars", k = 40, threshold = 10,
#         col = "ivory1", border = "ivory4", lwd = 0.6, add = TRUE)
#
#
#
#
#
#
# plot(bars$geometry, add=T, cex = .1)
# plot(iris)
#
# plot(irispt)
#
# plot(paris, add=T)
#
# library(sp)
# library(SpatialPosition)
# library(cartography)
#
# grid <- CreateGrid(w = paris, resolution = 200)
#
# plot(grid, add=F)
#
#
# par(mar = c(0,0,1.2,0))
# plot(grid, col = NA,  add=F)
# plot(paris, col = "grey50", border = NA, add=T)
# plot(grid, pch = 20, col = "black", cex = 0.3, add=T)
# layoutLayer(title = "Grille régulière d'un pas de 200 mètres",
#             frame = FALSE,
#             coltitle = "black",
#             col = NA, scale = NULL, sources="",
#             author = "")
#


R/package.R
#' @title Line Map Package
#' @rdname linemap-package
#' @name linemap-package
#' @description
#' Create maps made of lines. The package contains two functions:
#' \code{\link{linemap}} and \code{\link{getgrid}}.
#'
#' \code{\link{linemap}} displays a map made of lines using a data frame of
#' gridded data.
#'
#' \code{\link{getgrid}} transforms a set of polygons (\code{\link{sf}} objects)
#' into a suitable data frame
#' for \code{\link{linemap}}.
#' @note
#' These three mains sources gave me the inspiration to create
#' \code{linemap}:\cr
#' - Joy Division's 'Unknown Pleasures' Cover (\url{https://en.wikipedia.org/wiki/Unknown_Pleasures}) \cr
#' - the work of James Cheshire (Population Lines: How and Why I Created It - \url{https://jcheshire.com/featured-maps/population-lines-how-and-why-i-created-it/})\cr
#' - the work of Ryan Brideau (GeospatialLineGraphs - \url{https://github.com/Brideau/GeospatialLineGraphs})\cr
#' @docType package
NULL

#' @title Popultation Data for Occitanie
#' @name popOcc
#' @description Gridded population of Occitanie
#' @format data.frame
#' @field X longitude
#' @field Y latitude
#' @field pop population
#' @source Extract from INSEE's gridded population data, Population 2009(?) - \url{https://www.insee.fr/fr/statistiques/1405815}
#' @docType data
NULL




R/my_attempt.R
# library(foreign)
# France <- readRDS("France.rds")
# Occitanie <- readRDS("Occitanie.rds")
# reg <- readRDS("OccitanieReg.rds")
#
# df <- read.dbf("/home/tg/Téléchargements/ECP1KM_09_MET/R_rfl09_LAEA1000.dbf")
# df <- df[-1,]
#
#
# grid <- expand.grid(x=unique(df$x_laea), y = unique(df$y_laea))
#
# grid <- merge(grid, df[,c("x_laea","y_laea", "ind")],
#               by.x = c("x","y"),
#               by.y = c("x_laea","y_laea"), all.x = TRUE)
#
# grid[is.na(grid$ind),"ind"] <- 0
# gridx <- grid
#
# devtools::create_description()
#
#
# # grid <- gridx
# grid <- grid[grid$x>3425028 & grid$x <3917763 & grid$y>2159464 & grid$y <2482957,]
#
#
# lat <- unique(grid$y)
# lon <- unique(grid$x)
#
#
# par(mar=c(0,0,0,0))
# plot(1:10, xlim = bbox(France)[1,], ylim = bbox(France)[2,], asp=T)
#
# plot(reg, lwd = 0.7, border = NA, col = "ivory3", add=F)
# for (i in length(lat):1){
#   ly <- grid[grid$y==lat[i],]
#   ly$ind[ly$ind<50] <- 0
#   yVals <- ly$y + ly$ind * 3
#   xVals <- c(ly$x)
#   yVals[is.na(yVals)] <- lat[i]
#   yVals[1] <- lat[i]
#   yVals[length(yVals)] <- yVals[1]
#
#   polygon(xVals, yVals, border = NA, col = "#ffffff")
#   # lines(xVals, yVals, col="#8C8C8C", lwd=0.1)
#
#   j<- 1
#   while (j <= (length(yVals) - 1)) {
#
#     if ((ly$ind[j]) > 50 | (ly$ind[j+1]) > 50) {
#       segments(xVals[j], yVals[j], xVals[j+1], yVals[j+1], col="#000000", lwd=.5)
#     } else { } # Do nothing
#
#     j <- j + 1
#
#   }
# }
#
# # dev.off()
#
#
# # plot(reg, xlim = c(3425028, 3673294), ylim = c(2297453, 2369376))
# # plot(reg, lwd = 0.7, border = NA, col = "ivory3", add=F)
# #
# # for (i in length(lat):1){
# #   ly <- gridO[gridO$y==lat[i],]
# #  # points(ly$x, ly$y + ly$ind * 3, type = "l", lwd = 0.5, col = "black")
# #
# #
# #   xx <- c(ly$x, rev(ly$x))
# #   yy <- c(rep(lat[i], nrow(ly)), rev(ly$y + ly$ind * 3))
# #   yy[length(yy)] <- yy[1]
# #   yy[is.na(yy)] <- lat[i]
# #   pol = st_sfc(st_polygon(list(cbind(xx,yy))))
# #   polb <- st_buffer(st_buffer(pol, dist = -10), 10)
# #   plot(polb, add=T, border = "black", col = "ivory3", lwd = 0.5)
# #
# # }
#


R/myscdattempt.R
# library(sf)
# library(readxl)
# iris <- st_read(dsn = "/home/tg/Téléchargements/CONTOURS-IRIS_2-1__SHP_LAMB93_FXX_2016-11-10/CONTOURS-IRIS/1_DONNEES_LIVRAISON_2015/CONTOURS-IRIS_2-1_SHP_LAMB93_FE-2015/",layer = "CONTOURS-IRIS")
# xx <- read_excel(path = "/home/tg/Téléchargements/base-ic-activite-residents-2013.xls", sheet = "IRIS", skip = 5)
# dep <- c(29,22,56,35)
# iris75 <- iris[substr(iris$INSEE_COM,1,2) %in% dep,]
# x <- xx[xx$DEP %in% dep,]
# iris75 <- merge(iris75,x, by.x="CODE_IRIS",by.y = "IRIS", all.x=T)
#
#
# x <- iris75
# cellsize = 2000*2000
# var  = names(x)[19:120]
# getGridSquare <- function(x, cellsize){
#   # cellsize transform
#   cellsize = sqrt(cellsize)
#   boundingBox <- st_bbox(x)
#   rounder <- boundingBox[1:2] %% cellsize
#   boundingBox[1] <- boundingBox[1] - rounder[1]
#   boundingBox[2] <- boundingBox[2] - rounder[2]
#   n <- unname(c(ceiling(diff(boundingBox[c(1, 3)]) / cellsize),
#                 ceiling(diff(boundingBox[c(2, 4)]) / cellsize)))
#   grid <- st_make_grid(cellsize = cellsize, offset = boundingBox[1:2], n = n,
#                        crs = st_crs(x))
#   grid <- st_sf(id_cell=1:length(grid), geometry = grid)
#   # grid$id_cell <- 1:nrow(grid)
#   row.names(grid) <- grid$id_cell
#   return(grid)
# }
#
#
#
# x$area <- sf::st_area(x)
#
# # get a grid
# grid <- getGridSquare(x, cellsize)
# plot(grid$geometry, add=T)
#
#
#
# # keep only intersecting cells
# # gover <- sf::st_intersects(grid, x)
# # grid <- grid[unlist(lapply(gover, FUN = function(x) {if(length(x)>0){TRUE}else{FALSE}})), ]
#
#
# t0 <- Sys.time()
# # predicted warning, we don't care...
# options(warn = -1)
# parts <- sf::st_intersection(x = grid[,"id_cell"], y = x)
# options(warn = 0)
# t1 <- Sys.time()
# t1-t0
#
#
# parts$area_part <- sf::st_area(parts)
#
#
#
# lvar <- vector(mode = "list", length = length(var))
# names(lvar) <- var
# for (i in 1:length(lvar)){
#   lvar[[i]] <- as.vector(parts[[names(lvar)[i]]] * parts$area_part / parts$area)
# }
# v <- aggregate(do.call(cbind,lvar), by = list(id = parts[['id_cell']]),
#                FUN = sum, na.rm=TRUE)
#
# grid <- merge(grid, v, by.x  = "id_cell", by.y = "id", all.x = T)
#
#
#
# gr <- cbind(grid, st_coordinates(st_centroid(grid)))
#
#
#
#
# gr <- as.data.frame(gr[,c("X","Y",var), drop=TRUE])
# head(gr)
# library(linemap)
# plot(iris75$geometry, col = "ivory3", lwd = 0.2, border = NA)
# linemap(x = gr, var = "C13_ACT1564", k = 3, threshold = 100,col = "ivory1", border = "ivory4", lwd = 0.6, add = F)
#





# library(foreign)
# France <- readRDS("/home/tg/Documents/prj/linemap/France.rds")
# Occitanie <- readRDS("/home/tg/Documents/prj/linemap/Occitanie.rds")
# Region <- readRDS("/home/tg/Documents/prj/linemap/OccitanieReg.rds")
# df <- read.dbf("/home/tg/Téléchargements/ECP1KM_09_MET/R_rfl09_LAEA1000.dbf")
# df <- df[-1,]
# df <- df[df$x_laea>3425028 & df$x_laea <3917763 & df$y_laea>2159464 & df$y_laea <2482957,]
# grid <- expand.grid(x=unique(df$x_laea), y = unique(df$y_laea))
# grid <- merge(grid, df[,c("x_laea","y_laea", "ind")],
#               by.x = c("x","y"),
#               by.y = c("x_laea","y_laea"), all.x = TRUE)
#
# grid[is.na(grid$ind),"ind"] <- 0
# library(sf)
# grid = st_as_sf(grid, coords = c("x", "y"), crs = proj4string(Region), agr = "constant")
# xx <- st_intersects(x = grid, y = st_as_sf(Region))
# grid[unlist(lapply(xx, FUN = function(x) {if(length(x)>0){FALSE}else{TRUE}})), "ind"] <- 0
# Region <- st_as_sf(Region)
# x <- data.frame(st_coordinates(st_centroid(grid)), pop = grid$ind)
# str(Region)
# popOcc <- x
# regOcc <- st_simplify(Region, dTolerance = 1000)
# save(list=c("popOcc", "regOcc"), file = "data/Occitanie.RData" )

# dev.off()



R/smoothgrid.R
#' @title Smooth a Gridded Data Frame
#' @name smoothgrid
#' @description Smooth the grid using \link{focal} and \link{focalWeight} from \link{raster}.
#'
#' @param x a data.frame, two first column must be longitudes and latitudes of
#' gridded data.
#' @param var name of the variable to smooth
#' @param d numeric. If type=circle, the radius of the circle. If type=rectangle
#' the dimension of the rectangle (one or two numbers). If type=Gauss the size
#' of sigma, and optionally another number to determine the size of the matrix
#' returned (default is 3 times sigma)
#' @param type character indicating the type of filter to be returned
#'
#'
#' @return a smoother gridded data.frame
#' @export
#'
#' @examples
#' library(linemap)
#' data("popOcc")
#' # example on an extract of the gridded data
#' popTou <- popOcc[findInterval(popOcc$X, c(3600234,3659444)) == 1  &
#'                    findInterval(popOcc$Y, c(2290913,2348192)) == 1, ]
#' popTou1 <- smoothgrid(x = popTou, var = "pop", d = 1500, type = "circle")
#' popTou2 <- smoothgrid(x = popTou, var = "pop", d = 3000, type = "circle")
#' popTou3 <- smoothgrid(x = popTou, var = "pop", d = 5000, type = "circle")
#'
#' opar <- par(mar=c(0.2,0.2,0.2,0.2), bg = "ivory1", mfrow = c(1,4))
#'
#' linemap(x = popTou, var = "pop", k = 2.5, threshold = 50,
#'         col = "ivory1", border = "ivory4", lwd = 0.6, add = FALSE)
#' mtext(text = "No smoothing", side = 3, line = -1, adj = c(0.2))
#'
#' linemap(x = popTou1, var = "pop", k = 3, threshold = 50,
#'         col = "ivory1", border = "ivory4", lwd = 0.6, add = FALSE)
#' mtext(text = "type = 'circle', d = 1500", side = 3, line = -1,
#'       adj = c(0.2))
#'
#' linemap(x = popTou2, var = "pop", k = 4, threshold = 50,
#'         col = "ivory1", border = "ivory4", lwd = 0.6, add = FALSE)
#' mtext(text = "type = 'circle', d = 3000", side = 3, line = -1,
#'       adj = c(0.2))
#'
#' linemap(x = popTou3, var = "pop", k = 6, threshold = 50,
#'         col = "ivory1", border = "ivory4", lwd = 0.6, add = FALSE)
#' mtext(text = "type = 'circle', d = 5000", side = 3, line = -1,
#'       adj = c(0.2))
#'
#' par(opar)
#'
#' \donttest{
#' library(sf)
#' data("occitanie")
#' popOcc1 <- smoothgrid(x = popOcc, var = "pop", d = c(1000,5000), type = "Gauss")
#' opar <- par(mar=c(0,0,0,0), bg = "ivory2")
#' plot(st_geometry(occitanie), col="ivory1", border = NA)
#' linemap(x = popOcc1, var = "pop", k = 5, threshold = 25,
#'         col = "ivory1", border = "ivory4", lwd = 0.6, add = TRUE)
#' par(opar)
#' }
smoothgrid <- function(x, var, d, type){
  x <- sp::SpatialPointsDataFrame(coords = x[,1:2], data = x[, var, drop = FALSE])
  sp::gridded(x) <- TRUE
  x <- raster::raster(x)
  mat <- raster::focalWeight(x = x, d = d,type = type)
  x <- raster::focal(x = x, w = mat, fun = sum,
                     na.rm = FALSE, pad = TRUE, padValue = 0)
  result <- data.frame(sp::coordinates(x),
                       pop = as.vector(x, mode = "numeric"))
  names(result)[3] <- var
  return(result)
}
